deep_qa/data/vocabulary.py

```diff
@@ -0,0 +1,111 @@
+                self._index_to_token[namespace][i + 1] = token
+
+    def finalize(self):
+        logger.info("Finalizing data indexer")
+        self._finalized = True
+
+    def fit_token_dictionary(self, dataset, min_count: int=1):
+        """
+        Given a ``Dataset``, this method decides which tokens are given an index, and which ones are
+        mapped to an OOV token (in this case "UNK").  This method must be called before any dataset
+        is indexed with this ``Vocabulary``.  If you don't first fit the token dictionary, you'll
+        basically map every token onto "UNK".
+
+        We call ``instance.words()`` for each instance in the dataset, and then keep all tokens that
+        appear at least ``min_count`` times.
+
+        Parameters
+        ----------
+        dataset: ``TextDataset``
+            The dataset to index.
+
+        min_count: int, optional (default=1)
+            The minimum number of occurences a token must have in the dataset in order to be
+            assigned an index.
+        """
+        # TODO(matt): FIXME
+        logger.info("Fitting token dictionary with min count of %d, finalized is %s",
+                    min_count, self._finalized)
+        if self._finalized:
+            logger.warning("Trying to fit a finalized Vocabulary.  This is a no-op.  Did you "
+                           "really want to do this?")
+            return
+        namespace_token_counts = defaultdict(lambda: defaultdict(int))
+        for instance in tqdm.tqdm(dataset.instances):
+            namespace_dict = instance.words()
+            for namespace in namespace_dict:
+                for token in namespace_dict[namespace]:
+                    namespace_token_counts[namespace][token] += 1
+        for namespace in tqdm.tqdm(namespace_token_counts):
+            for token, count in namespace_token_counts[namespace].items():
+                if count >= min_count:
+                    self.add_token_to_namespace(token, namespace)
+
+    def add_token_to_namespace(self, token: str, namespace: str='tokens') -> int:
+        """
+        Adds ``token`` to the index, if it is not already present.  Either way, we return the index of
+        the token.
+        """
+        if self._finalized:
+            logger.warning("Trying to add a token to a finalized Vocabulary.  This is a no-op.  "
+                           "Did you really want to do this?")
+            return self._token_to_index[namespace].get(token, -1)
+        if token not in self._token_to_index[namespace]:
+            index = len(self._token_to_index[namespace])
+            self._token_to_index[namespace][token] = index
+            self._index_to_token[namespace][index] = token
+            return index
+        else:
+            return self._token_to_index[namespace][token]
+
+    def tokens_in_namespace(self, namespace: str='tokens'):
+        return self._token_to_index[namespace].keys()
```

Comment:
Hmm, looks like I need to switch `_index_to_token` to be a list, not a dictionary.